Replace 'ERR_EXPLAIN' with 'ERR_FAIL_*_MSG' in 'core/' and 'editor/'

Unholydeath opened the pull request:
I removed all the EXPLAIN error codes from /core, /drivers, and /editor except for a few files that I had troubles with: 
core/error_macros.h (defines EXPLAIN and uses it in the *_MSG calls, so don't know if it needs changed)
core/io/file_access_buffered_fa.h (couldn't find it in my files)
core/os/memory.cpp (combining CRASH_NOW with the EXPLAINC, seemed to only give errors)
drivers/png/png_driver_common.cpp (wasn't sure how to replace just a EXPLAINC)
drivers/xaudio2/audio_driver_xaudio2.cpp (couldn't find it in my files)

I also removed all the TTR() that I could find, which only existed in editor.

*Bugsquad edit:* Part of #31207.

KoBeWi commented:
You could remove conditions too. E.g. this
```
if (has_method(instance_type, mdname)) {
ERR_FAIL_V_MSG(NULL, "Class " + String(instance_type) + " already has a method " + String(mdname));
}
```
could become
```
ERR_FAIL_COND_V_MSG(has_method(instance_type, mdname), NULL, "Class " + String(instance_type) + " already has a method " + String(mdname));
```
If after condition there's only an `ERR_FAIL`, there's no reason it's not an `ERR_FAIL_COND`.

merumelu commented:
> I also removed all the TTR() that I could find, which only existed in editor.

You should only remove `TTR()` if it appears inside `ERR_EXPLAIN()`. Other errors are reported as part of the editor GUI and are supposed to be translatable.

akien-mga commented:
Indeed, sorry if my comment was not clear enough, but #31207 only covers `ERR_EXPLAIN()` calls and that's only there that TTR shouldn't be used (it hardly is a priori).

Unholydeath commented:
Ok, I'll try to revert some of the changes with TTR and find the ones that do matter.

Unholydeath commented:
I reverted my last commit that removed extra TTR(). I limited the changes to just the EXPLAIN()s that were changed to MSG()s. I also replaced a few if statements with the ERROR_*_MSG() commands.

KoBeWi commented:
> I also replaced a few if statements with the ERROR_*_MSG() commands.

I see some more that could be replaced, but that's probably not really important (might be done in another PR too if someone wants).

akien-mga commented:
There is some trailing whitespace to remove, see the diff in https://travis-ci.org/godotengine/godot/jobs/570311284 (select the text with the mouse to see whitespace on `-` lines).

The commits should be squashed together to remove the irrelevant temporary history about `TTR()` removals and the subsequent revert.

It would also be great if you could add dots at the end of error sentences, as discussed in https://github.com/godotengine/godot/issues/31207#issuecomment-519816715 and done in the other PRs.

Unholydeath commented:
I condensed many ifs and ERR_*_MSG statements, added dots to the end of *MSG sentences, and removed a few TTR() from *MSG() statements I missed before. I also removed extra white space and squashed the commits.

akien-mga commented on core/array.cpp:
```diff
@@ -134,16 +134,14 @@ void Array::erase(const Variant &p_value) {
 	if (_p->array.size() == 0) {
-		ERR_EXPLAIN("Can't take value from empty array");
-		ERR_FAIL_V(Variant());
+		ERR_FAIL_V_MSG(Variant(), "Can't take value from empty array.");
```
Same here.

akien-mga commented on core/bind/core_bind.cpp:
```diff
@@ -74,8 +74,7 @@ RES _ResourceLoader::load(const String &p_path, const String &p_type_hint, bool
 	if (err != OK) {
-		ERR_EXPLAIN("Error loading resource: '" + p_path + "'");
-		ERR_FAIL_V(ret);
+		ERR_FAIL_V_MSG(ret, "Error loading resource: '" + p_path + "'.");
```
Same here.

akien-mga commented on core/class_db.cpp:
```diff
@@ -924,16 +921,11 @@ void ClassDB::add_property(StringName p_class, const PropertyInfo &p_pinfo, cons
-			}
-		}
+
+		ERR_FAIL_COND_MSG(!mb_set, "Invalid Setter: " + p_class + "::" + p_setter + " for property: " + p_pinfo.name + ".");
```
"Setter" could be lowercase to match style used in most errors.

akien-mga commented on core/class_db.cpp:
```diff
@@ -943,25 +935,15 @@ void ClassDB::add_property(StringName p_class, const PropertyInfo &p_pinfo, cons
-			ERR_EXPLAIN("Invalid Getter: " + p_class + "::" + p_getter + " for property: " + p_pinfo.name);
-			ERR_FAIL();
-		} else {
+		ERR_FAIL_COND_MSG(!mb_get, "Invalid Getter: " + p_class + "::" + p_getter + " for property: " + p_pinfo.name + ".");
```
Lowercase getter.

akien-mga commented on core/color.cpp:
```diff
@@ -335,36 +335,23 @@ Color Color::html(const String &p_color) {
 	} else {
-		ERR_EXPLAIN("Invalid Color Code: " + p_color);
-		ERR_FAIL_V(Color());
+		ERR_FAIL_V_MSG(Color(), "Invalid Color Code: " + p_color + ".");
```
If you can also fix capitalization for the errors in this class, that would be nice. "Invalid color code: "

akien-mga commented on core/color.cpp:
```diff
@@ -428,8 +415,7 @@ Color Color::named(const String &p_name) {
 	} else {
-		ERR_EXPLAIN("Invalid Color Name: " + p_name);
-		ERR_FAIL_V(Color());
+		ERR_FAIL_V_MSG(Color(), "Invalid Color Name: " + p_name + ".");
```
The `if` / `else` could now be simplified as:
```cpp
ERR_FAIL_NULL_V_MSG(color, Color(), "Invalid Color Name: " + p_name + ".");

return color->value();
```
but that's not mandatory to do in this PR, the current code is fine too.

akien-mga commented on core/hash_map.h:
```diff
@@ -153,7 +153,7 @@ class HashMap {
 		if (!new_hash_table) {
 
-			ERR_PRINT("Out of Memory");
+			ERR_PRINT("Out of Memory.");
```
Please lowercase "Memory".

akien-mga replied:
This could actually be `ERR_FAIL_COND(!new_hash_table, "Out of memory.");`

akien-mga commented on core/hash_map.h:
```diff
@@ -499,8 +498,7 @@ class HashMap {
 			if (!e) {
-				ERR_EXPLAIN("Invalid key supplied")
-				ERR_FAIL_V(NULL);
+				ERR_FAIL_V_MSG(NULL, "Invalid key supplied.");
```
Same here.

akien-mga commented on core/io/ip_address.cpp:
```diff
@@ -81,8 +81,7 @@ static void _parse_hex(const String &p_string, int p_start, uint8_t *p_dst) {
 		} else {
-			ERR_EXPLAIN("Invalid character in ipv6 address: " + p_string);
-			ERR_FAIL();
+			ERR_FAIL_MSG("Invalid character in ipv6 address: " + p_string + ".");
```
Please fix capitalization of "IPv6".

akien-mga commented on core/io/ip_address.cpp:
```diff
@@ -166,10 +163,7 @@ void IP_Address::_parse_ipv4(const String &p_string, int p_start, uint8_t *p_ret
-		ERR_EXPLAIN("Invalid IP Address String: " + ip);
-		ERR_FAIL();
-	}
+	ERR_FAIL_COND_MSG(slices != 4, "Invalid IP Address String: " + ip + ".");
```
Capitalization, should be "Invalid IP address string: ".

akien-mga commented on core/io/marshalls.cpp:
```diff
@@ -379,8 +379,7 @@ Error decode_variant(Variant &r_variant, const uint8_t *p_buffer, int p_len, int
 
-			ERR_EXPLAIN("Can't marshallize resources");
-			ERR_FAIL_V(ERR_INVALID_DATA); //no, i'm sorry, no go
+			ERR_FAIL_V_MSG(ERR_INVALID_DATA, "Can't marshallize resources"); //no, i'm sorry, no go
```
That's actually dead code, there hasn't been `Variant::RESOURCE` in years, so you could remove the commented out code.

akien-mga commented on core/io/resource_format_binary.cpp:
```diff
@@ -711,16 +707,14 @@ Error ResourceInteractiveLoaderBinary::poll() {
-		ERR_EXPLAIN(local_path + ":Resource type in resource field not a resource, type is: " + obj->get_class());
 		memdelete(obj); //bye
-		ERR_FAIL_V(ERR_FILE_CORRUPT);
+		ERR_FAIL_V_MSG(ERR_FILE_CORRUPT, local_path + ":Resource type in resource field not a resource, type is: " + obj->get_class() + ".");
```
This is going to crash, you can dereference `obj` after deleting it. So you need to savev `obj->class()` in a temporary variable before freeing it.

akien-mga commented on core/io/resource_format_binary.cpp:
```diff
@@ -877,8 +870,7 @@ void ResourceInteractiveLoaderBinary::open(FileAccess *p_f) {
 		f->close();
-		ERR_EXPLAIN("File Format '" + itos(FORMAT_VERSION) + "." + itos(ver_major) + "." + itos(ver_minor) + "' is too new! Please upgrade to a new engine version: " + local_path);
-		ERR_FAIL();
+		ERR_FAIL_MSG("File Format '" + itos(FORMAT_VERSION) + "." + itos(ver_major) + "." + itos(ver_minor) + "' is too new! Please upgrade to a new engine version: " + local_path + ".");
```
Please lowercase "Format".

akien-mga commented on core/io/resource_format_binary.cpp:
```diff
@@ -926,8 +918,7 @@ void ResourceInteractiveLoaderBinary::open(FileAccess *p_f) {
 		error = ERR_FILE_CORRUPT;
-		ERR_EXPLAIN("Premature End Of File: " + local_path);
-		ERR_FAIL();
+		ERR_FAIL_MSG("Premature End Of File: " + local_path + ".");
```
Lowercase "End Of File" (but might be worth adding "EOF", e.g. "Premature end of file (EOF): ".

akien-mga commented on core/io/resource_format_binary.cpp:
```diff
@@ -1122,7 +1112,7 @@ Error ResourceFormatLoaderBinary::rename_dependencies(const String &p_path, cons
 		//use the old approach
 
-		WARN_PRINT(("This file is old, so it can't refactor dependencies, opening and resaving: " + p_path).utf8().get_data());
+		WARN_PRINT(("This file is old, so it can't refactor dependencies, opening and resaving: " + p_path + ".").utf8().get_data());
```
This could use `WARN_PRINTS` (the String version) to remove the need for the `.utf8().get_data()` C-string stuff.

akien-mga commented on core/io/resource_format_binary.cpp:
```diff
@@ -1153,8 +1143,7 @@ Error ResourceFormatLoaderBinary::rename_dependencies(const String &p_path, cons
 		memdelete(fw);
-		ERR_EXPLAIN("File Format '" + itos(FORMAT_VERSION) + "." + itos(ver_major) + "." + itos(ver_minor) + "' is too new! Please upgrade to a new engine version: " + local_path);
-		ERR_FAIL_V(ERR_FILE_UNRECOGNIZED);
+		ERR_FAIL_V_MSG(ERR_FILE_UNRECOGNIZED, "File Format '" + itos(FORMAT_VERSION) + "." + itos(ver_major) + "." + itos(ver_minor) + "' is too new! Please upgrade to a new engine version: " + local_path + ".");
```
Lowercase "Format".

akien-mga commented on core/io/resource_format_binary.cpp:
```diff
@@ -1497,8 +1486,7 @@ void ResourceFormatSaverBinaryInstance::write_variant(FileAccess *f, const Varia
 					f->store_32(OBJECT_EMPTY);
-					ERR_EXPLAIN("Resource was not pre cached for the resource section, most likely due to circular refedence.");
-					ERR_FAIL();
+					ERR_FAIL_MSG("Resource was not pre cached for the resource section, most likely due to circular refedence.");
```
Typo worth fixing: "refe**r**ence".

akien-mga commented on core/io/resource_loader.cpp:
```diff
@@ -395,8 +389,7 @@ RES ResourceLoader::load(const String &p_path, const String &p_type_hint, bool p
 		}
-		ERR_EXPLAIN("Remapping '" + local_path + "'failed.");
-		ERR_FAIL_V(RES());
+		ERR_FAIL_V_MSG(RES(), "Remapping '" + local_path + "'failed.");
```
Add missing space before "failed".

akien-mga commented on core/io/resource_loader.cpp:
```diff
@@ -480,8 +473,7 @@ Ref<ResourceInteractiveLoader> ResourceLoader::load_interactive(const String &p_
 		if (!success) {
-			ERR_EXPLAIN("Resource: '" + local_path + "' is already being loaded. Cyclic reference?");
-			ERR_FAIL_V(RES());
+			ERR_FAIL_V_MSG(RES(), "Resource: '" + local_path + "' is already being loaded. Cyclic reference?");
```
Should be `ERR_FAIL_COND_V_MSG(!success, ...)`.

akien-mga commented on core/io/resource_loader.cpp:
```diff
@@ -503,8 +495,7 @@ Ref<ResourceInteractiveLoader> ResourceLoader::load_interactive(const String &p_
 		}
-		ERR_EXPLAIN("Remapping '" + local_path + "'failed.");
-		ERR_FAIL_V(RES());
+		ERR_FAIL_V_MSG(RES(), "Remapping '" + local_path + "'failed.");
```
Also missing space.

akien-mga commented on core/math/octree.h:
```diff
@@ -568,10 +568,7 @@ void Octree<T, use_pairs, AL>::_ensure_valid_root(const AABB &p_aabb) {
-				ERR_EXPLAIN("Octree upper size limit reeached, does the AABB supplied contain NAN?");
-				ERR_FAIL();
-			}
+			ERR_FAIL_COND_MSG(base.size.x > OCTREE_SIZE_LIMIT, "Octree upper size limit reeached, does the AABB supplied contain NAN?");
```
Typo worth fixing in "r**ee**ached".

akien-mga commented on core/object.cpp:
```diff
@@ -709,20 +709,17 @@ static void _test_call_error(const StringName &p_func, const Variant::CallError
 
-			ERR_EXPLAIN("Error Calling Function: " + String(p_func) + " - Invalid type for argument " + itos(error.argument) + ", expected " + Variant::get_type_name(error.expected));
-			ERR_FAIL();
+			ERR_FAIL_MSG("Error Calling Function: " + String(p_func) + " - Invalid type for argument " + itos(error.argument) + ", expected " + Variant::get_type_name(error.expected) + ".");
```
Lowercase "Calling Function" in this class.

akien-mga commented on core/object.cpp:
```diff
@@ -1975,8 +1948,7 @@ Object::~Object() {
 		if (s->lock) {
-			ERR_EXPLAIN("Attempt to delete an object in the middle of a signal emission from it");
-			ERR_CONTINUE(s->lock > 0);
+			ERR_CONTINUE_MSG(s->lock > 0, "Attempt to delete an object in the middle of a signal emission from it.");
```
The `if` is redundant for the condition and could be removed.

akien-mga commented on core/project_settings.cpp:
```diff
@@ -647,8 +644,7 @@ Error ProjectSettings::_save_settings_binary(const String &p_file, const Map<Str
 
-		ERR_EXPLAIN("Couldn't save project.binary at " + p_file);
-		ERR_FAIL_COND_V(err, err);
+		ERR_FAIL_COND_V_MSG(err, err, "Couldn't save project.binary at " + p_file + ".");
```
The `if` can be removed, and I'd make the condition more explicit: `ERR_FAIL_COND_V_MSG(err != OK, err, ...)`.

akien-mga commented on core/project_settings.cpp:
```diff
@@ -738,10 +734,7 @@ Error ProjectSettings::_save_settings_text(const String &p_file, const Map<Strin
-		ERR_EXPLAIN("Couldn't save project.godot - " + p_file);
-		ERR_FAIL_COND_V(err, err);
-	}
+	ERR_FAIL_COND_V_MSG(err, err, "Couldn't save project.godot - " + p_file + ".");
```
Not very important, but I'd make the first `err` check more explicit: `err != OK`.

akien-mga commented on drivers/png/resource_saver_png.cpp:
```diff
@@ -39,8 +39,7 @@ Error ResourceSaverPNG::save(const String &p_path, const RES &p_resource, uint32
 
-	ERR_FAIL_COND_V(!texture.is_valid(), ERR_INVALID_PARAMETER);
-	ERR_EXPLAIN("Can't save empty texture as PNG");
+	ERR_FAIL_COND_V_MSG(!texture.is_valid(), ERR_INVALID_PARAMETER, "Can't save empty texture as PNG.");
```
That `ERR_EXPLAIN` was actually for the next condition.

akien-mga replied:
Fixed myself.

akien-mga commented on drivers/unix/ip_unix.cpp:
```diff
@@ -184,8 +184,7 @@ void IP_Unix::get_local_interfaces(Map<String, Interface_Info> *r_interfaces) co
-		ERR_EXPLAIN("Call to GetAdaptersAddresses failed with error " + itos(err));
-		ERR_FAIL();
+		ERR_FAIL_MSG("Call to GetAdaptersAddresses failed with error " + itos(err) + ".");
 		return;
```
That `return;` is dead code, should be removed.

akien-mga commented on drivers/windows/file_access_windows.cpp:
```diff
@@ -168,12 +168,10 @@ void FileAccessWindows::close() {
 			}
 
-			ERR_EXPLAIN("Safe save failed. This may be a permissions problem, but also may happen because you are running a paranoid antivirus. If this is the case, please switch to Windows Defender or disable the 'safe save' option in editor settings. This makes it work, but increases the risk of file corruption in a crash.");
+			ERR_FAIL_COND_MSG(rename_error, "Safe save failed. This may be a permissions problem, but also may happen because you are running a paranoid antivirus. If this is the case, please switch to Windows Defender or disable the 'safe save' option in editor settings. This makes it work, but increases the risk of file corruption in a crash.");
```
That's wrong, it will return earlier than before, and `save_path` will no longer be reset. The cond check is also redundant.

akien-mga commented on editor/doc/doc_data.cpp:
```diff
@@ -991,10 +985,8 @@ Error DocData::save_classes(const String &p_default_path, const Map<String, Stri
-			ERR_CONTINUE(err);
-		}
+
+		ERR_CONTINUE_MSG(err, "Can't write doc file: " + save_file + ".");
```
Please make the error cond more explicit with `err != OK`.

akien-mga commented on editor/editor_export.cpp:
```diff
@@ -1010,8 +1010,8 @@ Error EditorExportPlatform::save_pack(const Ref<EditorExportPreset> &p_preset, c
 		memdelete(f);
-		DirAccess::remove_file_or_error(tmppath);
-		ERR_FAIL_V_MSG(ERR_CANT_CREATE, "Can't open file to read from path: " + String(tmppath));
+    DirAccess::remove_file_or_error(tmppath);
```
Indentation issue.

akien-mga commented on editor/import/editor_import_collada.cpp:
```diff
@@ -1204,10 +1204,7 @@ Error ColladaImport::_create_resources(Collada::Node *p_node, bool p_use_compres
-					ERR_EXPLAIN("Controller Instance Source '" + ngsource + "' is neither skin or morph!");
-					ERR_FAIL_V(ERR_INVALID_DATA);
-				}
+				ERR_FAIL_COND_V_MSG(ngsource != "", ERR_INVALID_DATA, "Controller Instance Source '" + ngsource + "' is neither skin or morph!");
```
Lowercase "Instance Source".

akien-mga commented on editor/import/resource_importer_wav.cpp:
```diff
@@ -123,8 +123,7 @@ Error ResourceImporterWAV::import(const String &p_source_file, const String &p_s
 		memdelete(file);
-		ERR_EXPLAIN("Not a WAV file (no WAVE RIFF Header)");
-		ERR_FAIL_V(ERR_FILE_UNRECOGNIZED);
+		ERR_FAIL_V_MSG(ERR_FILE_UNRECOGNIZED, "Not a WAV file (no WAVE RIFF Header)");
```
Lowercase "Header" (and add dot).

akien-mga commented on editor/plugins/theme_editor_plugin.cpp:
```diff
@@ -189,11 +189,9 @@ void ThemeEditor::_save_template_cbk(String fname) {
-		ERR_EXPLAIN(TTR("Can't save theme to file:") + " " + filename);
-		return;
-	}
+	ERR_FAIL_COND_MSG(!file, "Can't save theme to file: " + filename + ".");
```
Just for the reference, the previous code was not printing the error, but it was a bug which is fixed here.

akien-mga commented on editor/scene_tree_dock.cpp:
```diff
@@ -986,9 +986,10 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 				if (!new_node) {
-					ERR_EXPLAIN("Creating root from favorite '" + selected_favorite_root + "' failed. Creating 'Node' instead.");
 					new_node = memnew(Node);
+					ERR_FAIL_COND_MSG(!new_node, "Creating root from favorite '" + selected_favorite_root + "' failed. Creating 'Node' instead.");
```
This adds a bug, the previous code did not fail. It set an error and then nothing was showing it (no `ERR_FAIL`), so what was intended is likely an `ERR_PRINTS`.

akien-mga commented:
Reviewed everything, great work!

I found a couple bugs, the rest of the comments are mostly proofreading of additional improvements that could be made on the changed code.

Once those comments are addressed, this should be good to merge.

Please run `git rebase origin/master` to remove the extraneous merge commit, and maybe amend the commit log of 090189da to remove references to stuff being reverted, which is not part of this commit anymore.

Finally, just FYI, the email you used to sign your commit (see https://github.com/godotengine/godot/commit/090189dad9164bcc4f46def684392435f83f3ae7.patch) is not known to GitHub, so it doesn't associate it with your "@Unholydeath" account. It's not a problem per se, but if you want to claim ownership of this commit and get a "Contributor" badge once it's merged, you'd have to add this email as secondary GitHub email, or change the email you use to sign the commit.

Unholydeath commented:
I removed the bugs, condensed the if and ERR_*_MSG, and fixed the error text you pointed out. I am seeing if I can squash it all again into 1 commit.

KoBeWi commented on core/color.cpp:
```diff
@@ -412,11 +412,7 @@ Color Color::named(const String &p_name) {
-	} else {
-		ERR_FAIL_V_MSG(Color(), "Invalid Color Name: " + p_name + ".");
-	}
+	ERR_FAIL_NULL_V_MSG(color, Color(), "Invalid color name: " + p_name + ".");
```
The removed `return color->value();` should be after the `ERR_FAIL`. Otherwise there's no correct return.

Unholydeath commented:
I saved off my files, reverted to an older commit, and then moved the files back in so I could squash the commits (any time I merged master into my branch I couldn't squash it).

akien-mga commented:
You still have an extra merge commit, and a lot of outdated messages in your main commit log:
![Screenshot_20190815_025925](https://user-images.githubusercontent.com/4701338/63066061-c0fd9a80-bf08-11e9-87cf-03c243cce077.png)

When you run `git rebase -i origin/master`, the merge commit should be removed. You can then `git commit --amend` the HEAD commit which should be the relevant one, and edit its log. Finally, update the PR with `git push --force`. See [PR workflow](http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase).

Unholydeath committed:
Replace 'ERR_EXPLAIN' with 'ERR_FAIL_*_MSG' in 'core/' and 'editor/'

Condensed some if and ERR statements. Added dots to end of error messages

Couldn't figure out EXPLAINC. These files gave me trouble: core/error_macros.h, core/io/file_access_buffered_fa.h (where is it?),
core/os/memory.cpp,
drivers/png/png_driver_common.cpp,
drivers/xaudio2/audio_driver_xaudio2.cpp (where is it?)

Unholydeath commented:
I am able to squash my commits, but if I try to push up, I always have a conflict with AppVeyor and the editor/editor_export.cpp file. If I resolve the conflict, it will add another item to be pushed up, so it won't be squashed into a single commit. Any suggestions?

qarmin commented:
after `git rebase` you should resolve conflicts in files
next you should run `git rebase --continue`
and at the end `git push -f`

This works for me, but I'm not exactly sure if that will works for you.
You should do a copy of your repo and try to do this commands

akien-mga commented:
> I downloaded git to perform the actions through the command line, but I don't know how to edit this to combine them together. I type git rebase -i origin/master but I'm not sure how to continue.

Simply save and quit the text that comes up, that's enough to remove the merge commit (since that's what `rebase` does). Then you can `git commit --amend` the commit log.

akien-mga commented:
I rebased and pushed a fixed version for you so that we can get this merged :)

Git rebase and solving merge conflicts can be a bit tricky, but keep trying, once you figure it out it's a relatively smooth workflow :)

akien-mga commented on core/hash_map.h:
```diff
@@ -209,8 +209,7 @@ class HashMap {
 		if (!e) {
-			ERR_EXPLAIN("Out of memory");
-			ERR_FAIL_V(NULL);
+			ERR_FAIL_V_MSG(NULL, "Out of memory.");
```
Fixed myself.

akien-mga commented:
Thanks!
